### ISSUE TITLE
Mobile keyboard doesn't include  `/` char when expiration field is in focus

### ISSUE THREAD

**Issue by Julesssss** [Expensify/App#13133]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

CC @Gonals 

## Action Performed:
- Navigate to Profile > Payments > Add A debit Card
- Fill in the data until you get to the expiration date field
- Select the expiration date input 
- Enter 2 characters

## Expected Result:
- Either the `/` char should appear in the soft keyboard, or we should automatically add it for users

## Actual Result:
- Neither the `/` char appears in the soft keyboard, or do we automatically add it for users

![Screenshot_20221129_152213](https://user-images.githubusercontent.com/10736861/204571272-f951974c-0166-49b8-b3c1-31b6773562e0.png)

_Note: Users still seem to pass validation without it, but this is still confusing_

## Platform:

Where is this issue occurring?

- iOS
- Android
- Mobile Web

**Version Number:** all
**Reproducible in staging?:**  yes
**Reproducible in production?:**  yes
**Email or phone of affected tester (no customers):** all

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Puneet-here** [comment]
### Proposal


I suggest that we change the placeholder to `MMYY` so that the user will know that they don't have to add `/` Many sites do the same.
https://github.com/Expensify/App/blob/3e47bce7f7d98812735ba9b00de282a4fa9345ed/src/languages/en.js#L415
Another solution is to remove the keyboard this way the user can switch the keyboard to add numbers and `/`
We have keyboard type `numbers-and-punctuation` but it is only for ios. If we change this to `numbers-and-punctuation` this won't be a problem in ios and in android user will see default keyboard 
https://github.com/Expensify/App/blob/3e47bce7f7d98812735ba9b00de282a4fa9345ed/src/pages/settings/Payments/AddDebitCardPage.js#L141

**Comment by mananjadhav** [comment]
@Julesssss is this about not able to add from keyboard? or not able to add slash at all? I remember working on this PR [here](https://github.com/Expensify/App/issues/6891) to add auto-slash

**Comment by Julesssss** [comment]
> @Julesssss is this about not able to add from keyboard? or not able to add slash at all? I remember working on this PR [here](https://github.com/Expensify/App/issues/6891) to add auto-slash

Not during my testing today. We use the numerical keyboard type, and it wasn't auto-added. I do also recall a conversation about this though 😕

**Comment by Puneet-here** [comment]
It was removed [here](https://github.com/Expensify/App/pull/9056/commits/100ebd0152d645bc23f1526c7afe2ce84f9d0bf9). We can add it back.

**Comment by kadiealexander** [comment]
From this SO: https://stackoverflow.com/c/expensify/questions/14418
- [ ] **The "bug" is actually a bug**: Yes!
- [ ] **The bug is not a duplicate report of an existing GH (close the GH and add any novel details to the original GH instead)**: No
- [ ] **The bug is reproducible, following the reproduction steps:** Yes
![6D41C19D-8DFE-4E5F-AC51-AFEBEAD7F1AA](https://user-images.githubusercontent.com/59587260/204716426-6dd3cddd-060e-4977-9991-5c10dcc683be.png)
- [x] **If you’re unable to reproduce the bug, add the _Needs reproduction_ label. Comment on the issue outlining the steps you took to try to reproduce the bug, your results and tag the issue reporter and the Applause QA member who created the issue. Ask them to clarify reproduction steps and/or to check the reproduction steps again. Close issue.**: n/a
- [ ] **The GH template is filled out as fully as possible -- this means the GH body and title are clear (ie. could an external contributor understand it and work on it?):** Yes
- [ ] **The GH was created by an Expensify employee or a QA tester:** Yes
- [ ] **If the bug is an OldDot issue, you should decide whether it is widespread enough to justify fixing or it is better to wait for reunification. If it’s better to wait, close the GH & provide this justification:** n/a
- [ ] **If there's a link to Slack, check the discussion to see if we decided not to fix it:** n/a
- [ ] **Decide if the GH should be resolved by an External contributor or Internal engineer, add the appropriate label:** External.

**Comment by thesahindia** [comment]
> It was removed [here](https://github.com/Expensify/App/pull/9056/commits/100ebd0152d645bc23f1526c7afe2ce84f9d0bf9). ~We can add it back.~
> Looks like we can't add it back because of `Form` component

@Julesssss, any thoughts on this? We can't interfere with the value right now. We need to wait and see if someone can refactor the `Form` component to do so.

**Comment by railway17** [comment]
### Proposal

1. Reproducible
This issue is easy to reproduce and I don't think I have to attach video or screenshot again. 😄 
2. Why reproduced?
As @Puneet-here mentioned, appending slash module was removed while implementing with new `Form` logic.
And need to update the our `Form` as @thesahindia mentioned.
3. How to resolve?
- Add new props `shouldAppendSlash` props to the `TextInput` in `AddDebitCartPage.js`
```diff
        <TextInput
                                inputID="expirationDate"
                                label={this.props.translate('addDebitCardPage.expiration')}
                                placeholder={this.props.translate('addDebitCardPage.expirationDate')}
                                keyboardType={CONST.KEYBOARD_TYPE.NUMBER_PAD}
+                                shouldAppendSlash
                            />
```

- Update the `Form` component like below:
```diff
        this.inputRefs = {};
        this.touchedInputs = {};
+        this.allowSplashAutoAddable = false;

        this.setTouchedInput = this.setTouchedInput.bind(this);
....
                 onInputChange: (value, key) => {
                     const inputKey = key || inputID;
-                    this.setState(prevState => ({
-                        inputValues: {
-                            ...prevState.inputValues,
-                            [inputKey]: value,
-                        },
-                    }), () => this.validate(this.state.inputValues));
+                    if (child.props.shouldAppendSlash && this.allowSplashAutoAddable) {
+                        let slashAppended = ''; 
+                        if (value.length >= 2 && value.indexOf('/') == -1) {
+                            slashAppended += value.substring(0, 2) + '/' + value.substring(2);
+                        } else {
+                            slashAppended = value;
+                        }
+                        this.setState(prevState => ({
+                            inputValues: {
+                                ...prevState.inputValues,
+                                [inputKey]: slashAppended,
+                            },
+                        }), () => this.validate(this.state.inputValues));                        
+                    } else {
+                        this.setState(prevState => ({
+                            inputValues: {
+                                ...prevState.inputValues,
+                                [inputKey]: value,
+                            },
+                        }), () => this.validate(this.state.inputValues));
+                    }
...
                          child.props.onValueChange(value);
                    }
                },
+                onKeyPress: ({nativeEvent}) => {
+                    if (child.props.shouldAppendSlash) {
+                        if (nativeEvent.key == 'Backspace' && this.state.inputValues[inputID].length == 3) {
+                           this.allowSplashAutoAddable = false;
+                        } else {
+                            this.allowSplashAutoAddable = true;
+                        }
+                    }
+                }
+            });
        });
    }

```

https://user-images.githubusercontent.com/29966461/204745809-efc937bd-4147-47c1-b0da-d4a9586c2b51.mov

**Comment by varshamb** [comment]
### Proposal

We can use `keyboardType='numeric'` specific for Android.

https://github.com/Expensify/App/blob/3e47bce7f7d98812735ba9b00de282a4fa9345ed/src/pages/settings/Payments/AddDebitCardPage.js#L141

**Comment by thesahindia** [comment]
@railway17, I don't think that's a good solution. We want to keep the logic at AddDebitCardPage.js.

**Comment by thesahindia** [comment]
@varshamb, how is that going to solve this issue?

**Comment by bernhardoj** [comment]
### **Proposal**

We can receive a new `prop` for the Form children, let say it is called `transformer`, which will receive a function with 2 parameter that will transform the input every input change, in our case appending a `/`.

```diff
<TextInput
    inputID="expirationDate"
    label={this.props.translate('addDebitCardPage.expiration')}
    placeholder={this.props.translate('addDebitCardPage.expirationDate')}
    keyboardType={CONST.KEYBOARD_TYPE.NUMBER_PAD}
+   transformer={this.addOrRemoveSlashToExpiryDate}
/>
```
The `addOrRemoveSlashToExpiryDate` taken from old code with few changes.

```
    addOrRemoveSlashToExpiryDate(prevExpiryDate = '', inputExpiryDate) {
        let expiryDate = inputExpiryDate;

        // Remove the digit and '/' when backspace is pressed with expiry date ending with '/'
        if (inputExpiryDate.length < prevExpiryDate.length
            && (((inputExpiryDate.length === 3 && lodashEndsWith(inputExpiryDate, '/'))
                || (inputExpiryDate.length === 2 && lodashEndsWith(prevExpiryDate, '/'))))) {
            expiryDate = inputExpiryDate.substring(0, inputExpiryDate.length - 1);
        } else if (inputExpiryDate.length === 2 && _.indexOf(inputExpiryDate, '/') === -1) {
            // An Expiry Date was added, so we should append a slash '/'
            expiryDate = `${inputExpiryDate}/`;
        } else if (inputExpiryDate.length > 2 && _.indexOf(inputExpiryDate, '/') === -1) {
            // Expiry Date with MM and YY without slash, hence adding slash(/)
            expiryDate = `${inputExpiryDate.slice(0, 2)}/${inputExpiryDate.slice(2)}`;
        }
        return expiryDate
    }
```

And we will use that function on Form children `onInputChange` props

```diff
onInputChange: (value, key) => {
    const inputKey = key || inputID;
    this.setState(prevState => ({
        inputValues: {
            ...prevState.inputValues,
-           [inputKey]: value,
+           [inputKey]: child.props.transformer?.(prevState.inputValues[inputKey], value) ?? value,
        },
     }), () => this.validate(this.state.inputValues));
     ....
}
```

**Result**

https://user-images.githubusercontent.com/50919443/204765372-f609904a-c82d-414f-b6c5-2eae57f5cdc8.mp4

**Comment by thesahindia** [comment]
That looks great @bernhardoj! I need to test all the cases first since it can cause regression. It's EOD for me I will test this in the morning. Meanwhile can you update the proposal to resolve eslint errors?

**Comment by thesahindia** [comment]
Also cc: @Julesssss. Should we get some opinions on slack ? Concerned because of this [part] (https://github.com/Expensify/App/blob/main/contributingGuides/FORMS.md#modifying-user-input-on-change).

**Comment by bernhardoj** [comment]
Sorry, I am not aware there is some eslint errors. I have updated the code. @thesahindia

**Comment by thesahindia** [comment]
> Also cc: @Julesssss. Should we get some opinions on slack ? Concerned because of this [part](https://github.com/Expensify/App/blob/main/contributingGuides/FORMS.md#modifying-user-input-on-change).

Gentle bump @Julesssss

**Comment by Julesssss** [comment]
Yeah, given that this has been discussed before we should probably discuss further. I'm going to hold this in the meantime.

**Comment by thesahindia** [comment]
And it's monday... Started the [discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1670229793085299?thread_ts=1670229793.085299&cid=C01GTK53T8Q)

**Comment by lordthien** [comment]
**Proposal**
in file `AddDebitCartPage.js`
1. Add state `expirationDate` inside constructor
```diff
    constructor(props) {
        super(props);

        this.validate = this.validate.bind(this);
+        this.state = { expirationDate: "" };
        PaymentMethods.clearDebitCardFormErrorAndSubmit();
    }
```
2. Add function `handleTextExpirationDate`:
```diff
    handleTextExpirationDate(text) {
        let textTemp = text;
        if (textTemp[0] !== '1' && textTemp[0] !== '0') {
            textTemp = '';
        }
        if (textTemp.length === 2) {
            if (parseInt(textTemp.substring(0, 2)) > 12 || parseInt(textTemp.substring(0, 2)) == 0) {
                textTemp = textTemp[0];
            } else if (textTemp.length === 2 && this.state.expirationDate.length != 3) {
                textTemp += "/";
            } else if (text.length == 2 && this.state.expirationDate.length == 3) {
                textTemp = text.substring(0, text.length - 1)
            }
            else {
                textTemp = textTemp[0];
            }
        }
        this.setState({ expirationDate: textTemp })
    }
```
4. Add `maxLength={5}` (optional) & Update `TextInput`:
```diff
                            <TextInput
                                inputID="expirationDate"
                                label={this.props.translate('addDebitCardPage.expiration')}
                                placeholder={this.props.translate('addDebitCardPage.expirationDate')}
                                keyboardType={CONST.KEYBOARD_TYPE.NUMBER_PAD}
+                               maxLength={5}
+                               value={this.state.expirationDate}
+                               onChangeText={(text) => { this.handleTextExpirationDate(text) }}
                            />
```
![abnv](https://user-images.githubusercontent.com/71846803/205594622-54ad9efc-f681-4d05-ad85-a1e3033352d4.gif)

**Comment by thesahindia** [comment]
@lordthien, it's on hold right now.

**Comment by lordthien** [comment]
> @lordthien, it's on hold right now.

When will this issue re-start?

**Comment by Julesssss** [comment]
> When will this issue re-start?

When we have [agreed](https://expensify.slack.com/archives/C01GTK53T8Q/p1670229793085299) on what the desired behavior should be.

**Comment by Julesssss** [comment]
@thesahindia I'm happy with that, but we should also remove the ability to enter 5 characters.

**Comment by Puneet-here** [comment]
@Julesssss, are we waiting to discuss about character limit?

**Comment by Puneet-here** [comment]
### Updated proposal

Change the copy to `MMYY`
https://github.com/Expensify/App/blob/498e23273ff4c16431a3aa29229ca6aacd6ece41/src/languages/en.js#L421
Change the copy to `MMAA`
https://github.com/Expensify/App/blob/498e23273ff4c16431a3aa29229ca6aacd6ece41/src/languages/es.js#L421
Pass maxLength={4} below
https://github.com/Expensify/App/blob/498e23273ff4c16431a3aa29229ca6aacd6ece41/src/pages/settings/Payments/AddDebitCardPage.js#L137

cc: @Julesssss

**Comment by Puneet-here** [comment]
@Julesssss, I am seeing that you changes the MMAA to MMYY for the spanish copy too, I think it should be `MMAA`

**Comment by Julesssss** [comment]
Oh sorry, I can't read. That makes sense, thanks!

**Comment by kadiealexander** [comment]
Paid @thesahindia and @Puneet-here the price + bonus for this issue.

**Comment by Julesssss** [comment]
^ This was not a regression

**Comment by mallenexpensify** [comment]
> ^ This was not a regression

@Julesssss , we still want BZ to update the regression steps in TestRail so QA can check, right?

**Comment by thesahindia** [comment]
No actions are needed from me so unassigning.

**Comment by kadiealexander** [comment]
@Julesssss, bumping to get your thoughts on https://github.com/Expensify/App/issues/13133#issuecomment-1364431534

**Comment by kadiealexander** [comment]
Jules is on leave until next week.

**Comment by kadiealexander** [comment]
Not overdue.

**Comment by Julesssss** [comment]
Hey, sorry but I ran out of time for this today. Will return to this tomorrow.

**Comment by Julesssss** [comment]
> @Julesssss , we still want BZ to update the regression steps in TestRail so QA can check, right?

I don't think so in this case. This was specifically designed this way originally, and we simpy improved it.
